stratosphere/loader/source/ldr_capabilities.cpp

```diff
@@ -0,0 +1,428 @@
+
+            bool IsValid(const u32 *kac, size_t kac_count) const {
+                for (size_t i = 0; i < kac_count; i++) {
+                    if (GetCapabilityId(kac[i]) == this->GetId()) {
+                        const auto restrict = Decode(kac[i]);
+
+                        if (this->GetMinimumThreadPriority() < restrict.GetMinimumThreadPriority() ||
+                            this->GetMaximumThreadPriority() > restrict.GetMaximumThreadPriority() ||
+                            this->GetMinimumThreadPriority() > this->GetMaximumThreadPriority()) {
+                            return false;
+                        }
+
+                        if (this->GetMinimumCoreId() < restrict.GetMinimumCoreId() ||
+                            this->GetMaximumCoreId() > restrict.GetMaximumCoreId() ||
+                            this->GetMinimumCoreId() > this->GetMaximumCoreId()) {
+                            return false;
+                        }
+
+                        return true;
+                    }
+                }
+                return false;
+            }
+        );
+
+        DEFINE_CAPABILITY_CLASS(SyscallMask,
+            u32 GetMask() const {
+                return (this->GetValue() >> 0) & 0xFFFFFF;
+            }
+
+            u32 GetIndex() const {
+                return (this->GetValue() >> 24) & 0x7;
+            }
+
+            bool IsValid(const u32 *kac, size_t kac_count) const {
+                for (size_t i = 0; i < kac_count; i++) {
+                    if (GetCapabilityId(kac[i]) == this->GetId()) {
+                        const auto restrict = Decode(kac[i]);
+
+                        if (this->GetIndex() == restrict.GetIndex() && this->GetMask() == restrict.GetMask()) {
+                            return true;
+                        }
+                    }
+                }
+                return false;
+            }
+        );
+
+        DEFINE_CAPABILITY_CLASS(MapRange,
+            static constexpr size_t SizeMax = 0x100000;
+
+            u32 GetAddressSize() const {
+                return (this->GetValue() >> 0) & 0xFFFFFF;
+            }
+
+            u32 GetFlag() const {
+                return (this->GetValue() >> 24) & 0x1;
+            }
+
+            bool IsValid(const u32 next_cap, const u32 *kac, size_t kac_count) const {
+                if (GetCapabilityId(next_cap) != this->GetId()) {
+                    return false;
+                }
+
+                const auto next = Decode(next_cap);
+                const u32 start = this->GetAddressSize();
+                const u32 size = next.GetAddressSize();
+                const u32 end = start + size;
+                if (size >= SizeMax) {
+                    return false;
+                }
+
+                for (size_t i = 0; i < kac_count; i++) {
+                    if (GetCapabilityId(kac[i]) == this->GetId()) {
+                        const auto restrict = Decode(kac[i]);
+                        i++;
```

Comment:
Could move this up so that you have `kac[i++]`.

Reply:
Nice.